### ISSUE TITLE
nosp branch generates cgetpccsetoffset        $c12, test

### ISSUE THREAD

**Issue by RichardsonAlex** [CTSRD-CHERI/llvm#250]
Test case is derived from cheri-vla.c in clang, and generates the following binary output:
```
      a4:       48 0c 01 ff     cgetpccsetoffset        $c12, $zero
                00000000000000a4:  R_MIPS_32/R_MIPS_NONE/R_MIPS_NONE    test
```

**Comment by davidchisnall** [comment]
The confusing thing for me is why we didn't see this bug prior to the nosp branch, but I suspect that it's because we never tried generating non-PIC pure-cap code, which is the only situation where this would appear.  If you add -relocation-model=pic to the command, then the bug goes away.

It appears that, in non-PIC code generation, we create a DAG node for a `llvm.cheri.cap.offset.set` intrinsic, which has a TargetGlobal as the argument.  This then gets folded into the `CGetPCCSetOffset` instruction because its base is `CGetPCC`.

**Comment by RichardsonAlex** [comment]
Ah that makes sense, so it is probably a bug in master as well. For some reason I thought PIC was default for CHERI purecap when using llc as well not just clang.

**Comment by davidchisnall** [comment]
Indeed, it looks as if MipsISelLowering.cpp:3380 just wraps the DAG node in a get-pcc, set-offset sequence.  This works fine for PIC, because the node is a load (which always ends up in a register), but is a problem for non-PIC, because the node is a `TargetGlobalAddress`, which is expanded later.  I'm not sure what the best solution is here: probably to make sure that the expansion of TargetGlobalAddress in pure-cap mode always ends up in a register?
